### DISCUSSION TITLE
Usage of images from docker.io

### DISCUSSION THREAD

**Discussion by roee88** [fybrik/fybrik#260] [Developers] [answered]
The CI started to fail pulling images from docker.io (used e.g. by base of secret-provider, registry, etc.). I'm not sure if this is something that we did or if we just hit the recently added rate limits in docker.io. 

If this isn't a bug, do we want to vendor or find alternative base images? Maybe just add docker.io login to increase the limit?

@hunchback

**Comment by hunchback** [comment]
https://www.docker.com/increase-rate-limits from it i understand we are limited to 100 downloads / 6 hours which can explain why after a couple of builds we reach the limited (as we build 10 images).

we can consider a paid account
moving to a different repository
reducing number of docker image builds
caching base images in travis

**Comment by roee88** [comment]
From https://docs.travis-ci.com/user/docker/:
> Please note that, due to the upcoming Docker Rate Limit announcement, users will be required to add their own authentication information to their build settings or build config as documented below.

It might be worse than what you described as it seems to be shared between Travis users. Unless we completely eliminate using docker.io images, we must open an account (not necessarily a paid account though). After that, caching can help us not reach the personal limit, e.g. we can configure [mirror.gcr.io](https://cloud.google.com/container-registry/docs/pulling-cached-images).

\+ @ronenkat

**Comment by revit13** [comment]
I got the  " reached your pull rate limit." when running make run-integration-tests:

Taken from vault pod describe:
 Failed to pull image "banzaicloud/bank-vaults:master": rpc error: code = Unknown desc = failed to pull and unpack image "docker.io/banzaicloud/bank-vaults:master": failed to copy: httpReaderSeeker: failed open: unexpected status code https://registry-1.docker.io/v2/banzaicloud/bank-vaults/manifests/sha256:e0f29c1b63dd27afbc41d622fd84ef1a79785dbf8ff84e7b884c688a12a35c49: 429 Too Many Requests - Server message: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit

**Reply by ronenkat** [reply]
@revit13 is this for running integration-test one time? or more?
I believe that you are using anonymous access, and I assume that each IP is an anonymous user of its own.

**Reply by revit13** [reply]
@ronenkat I ran it several times.

**Reply by revit13** [reply]
@ronenkat I tried from a different VM and I get the same error. Its a new VM and I'm getting:

sudo docker run hello-world
Unable to find image 'hello-world:latest' locally
docker: Error response from daemon: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit.

**Reply by roee88** [reply]
Maybe it's based on the IP address and all these VMs go to dockethub from the same IP. This is where authentication (even not a paid account) could help.

**Comment by roee88** [comment]
Can we please list the images that we take from docker.io?

Here is what I got so far:
- openjdk:8u131-jre-alpine is used by Egeria connector
- python:3.6-slim used by secret manager
- golang:1.13.8-alpine used by test/services (use distroless instead)
- golang:1.13 used by policy compiler (old leftover??)
- node:argon used by sample gui (not used in CI)
- vault images from banzaicloud (to be replaced by official vault images but they are also on dockerhub)

What else? This is an issue both for CI and for end users.

**Reply by hunchback** [reply]
so for building purposes:

```
python                                                    3.6-slim            29b8dbd5cf2f        5 weeks ago         111MB
python                                                    3.6                 1297140c6dd2        2 months ago        874MB
alpine                                                    latest              d6e46aa2470d        2 months ago        5.57MB
registry                                                  2                   2d4f4b5309b1        6 months ago        26.2MB
kindest/node                                              v1.16.9             f74990728e72        8 months ago        1.25GB
openjdk                                                   8u131-jre-alpine    d5b43d97f41b        3 years ago         81.4MB
```

**Reply by hunchback** [reply]
for k8s deployment purposes taken postivly from docker.io:

```
      1 docker.io/banzaicloud/bank-vaults:master
      2 docker.io/kindest/kindnetd:0.5.4
      1 docker.io/library/vault:1.3.1
      1 docker.io/openpolicyagent/opa:latest
      1 docker.io/prom/statsd-exporter:latest
      1 docker.io/rancher/local-path-provisioner:v0.0.12
```

the following may be taken from docker.io:

```
      2 kindest/kindnetd:0.5.4
      1 openpolicyagent/opa:latest
      1 prom/statsd-exporter:latest
      1 rancher/local-path-provisioner:v0.0.12
      1 vault:1.3.1
```

this is taken using this list after running

```
make run-deploy-tests
```

```
aidan@AIDANSHRIBMAN-SDFUE6N:~/projects/the-mesh-for-data$ kubectl get pods --all-namespaces -o jsonpath="{..image}" |tr -s '[[:space:]]' '\n' |sort |uniq -c
      2 banzaicloud/bank-vaults:master
      2 docker.io/banzaicloud/bank-vaults:master
      2 docker.io/kindest/kindnetd:0.5.4
      1 docker.io/library/vault:1.3.1
      1 docker.io/openpolicyagent/opa:latest
      1 docker.io/prom/statsd-exporter:latest
      1 docker.io/rancher/local-path-provisioner:v0.0.12
      2 ghcr.io/banzaicloud/vault-operator:1.9.0
      2 ghcr.io/the-mesh-for-data/egr-connector:latest
      2 ghcr.io/the-mesh-for-data/opa-connector:latest
      2 ghcr.io/the-mesh-for-data/vault-connector:latest
      4 k8s.gcr.io/coredns:1.6.2
      2 k8s.gcr.io/etcd:3.3.15-0
      2 k8s.gcr.io/kube-apiserver:v1.16.9
      2 k8s.gcr.io/kube-controller-manager:v1.16.9
      4 k8s.gcr.io/kube-proxy:v1.16.9
      2 k8s.gcr.io/kube-scheduler:v1.16.9
      2 kindest/kindnetd:0.5.4
      1 openpolicyagent/opa:latest
      1 prom/statsd-exporter:latest
      2 quay.io/jetstack/cert-manager-cainjector:v0.13.1
      2 quay.io/jetstack/cert-manager-controller:v0.13.1
      2 quay.io/jetstack/cert-manager-webhook:v0.13.1
      1 rancher/local-path-provisioner:v0.0.12
      1 vault:1.3.1
aidan@AIDANSHRIBMAN-SDFUE6N:~/projects/the-mesh-for-data$
```

**Comment by ronenkat** [comment]
@roee88 @hunchback can https://www.docker.com/blog/expanded-support-for-open-source-software-projects/  help us?

**Reply by hunchback** [reply]
> 
> 
> @roee88 @hunchback can https://www.docker.com/blog/expanded-support-for-open-source-software-projects/ help us?

looking at this now

**Reply by hunchback** [reply]
I think this may be restrictive, as they require that we participate in promotional content something which I think IBM legal may frown upon

**Reply by ronenkat** [reply]
ok. @hunchback would the mirror help? see https://docs.docker.com/registry/recipes/mirror/  
At least for users who wants to build.
CI may be a different story.

**Reply by roee88** [reply]
What mirror did you have in mind? I linked to gcp mirror earlier but it isn't a real solution (no guarantee that an image will be in the mirror, especially if not using latest tag).

**Reply by roee88** [reply]
If we create our own mirror for the images we use it could be a solution for the CI.

**Reply by hunchback** [reply]
but a mirror is infrastructure we need to maintain ourselves ... which is less favorable. When I work locally from my laptop things appears to work usually. @revit13 where are you working from? (from your laptop or from a VM in HRL) ?

**Reply by revit13** [reply]
@hunchback I am working on VM.

**Comment by roee88** [comment]
So @ronenkat will open a free account using the-mesh-for-data user. 

Then @hunchback if you can please add the login step to the CI so it would login to dockerhub using that account. This will allow us to have 200 pulls per 6 hours rather than the ~0 we get today. Can you open an issue for it? Ronen will probably share with you the user details privately.

Moving forward we will need to lower the chance of hitting the limit by:
1. Using the gcp mirror in the CI
1. Lowering the number of images we take from dockerhub to minimum (thanks @hunchback for the detailed list)

And of course adding docs for end users. Fun.

**Reply by ronenkat** [reply]
@roee88  @hunchback Opened themeshfordata account on DockerHub.

**Comment by hunchback** [comment]
> @roee88 - Then @hunchback if you can please add the login step to the CI so it would login to dockerhub using that account. This will allow us to have 200 pulls per 6 hours rather than the ~0 we get today. Can you open an issue for it? Ronen will probably share with you the user details privately.

I suggest we first check how many pulls we are actually getting now (before running to fix the issue) as we are not getting 0 for sure - I suggest I create a travis job to count this (obviously this will eat all our allowance for 6 hours) - please let me know if you are ok with this

**Reply by roee88** [reply]
> Please note that, due to the upcoming Docker Rate Limit announcement, users will be required to add their own authentication information to their build settings or build config as documented below.

You are required to do that anyway. Counting won't help you, it depends on what *other Travis users* do in parallel.

**Comment by hunchback** [comment]
@roee88 / @ronenkat I suggest the following alternative, that we create a copy of any image we use on github registry and then change our: 
(1) docker build
(2) k8s deployment
to use these ... this will provide a long term solution for both CI and local build with a smaller investment

**Reply by roee88** [reply]
If you use an image that its base image is in dockerhub then it's more difficult. I'm all in as a next step as described in https://github.com/IBM/the-mesh-for-data/discussions/260#discussioncomment-272982.

**Reply by roee88** [reply]
Discussed externally and @hunchback pointed out that the login credentials won't be available in PR builds so it won't help us.

**Comment by roee88** [comment, marked as answer]
We're getting help with getting the base images mirrored in quay.io/ibm and switching to these as base images. Issue https://github.com/IBM/the-mesh-for-data/issues/274 opened.
